Changes to restore ipsec show command

bradyallenjohnson opened the pull request:
- Now called "ipsec briefconnectionstatus", and is a simpler version "ipsec connectionstatus"
- Now implemented in c-code in the whack binary

cagney commented:
It looks like the output is the same as:
```
ipsec connectionstatus | grep '[.][.][.]'
```

I wonder what we would think of displaying if the command was called `ipsec routestatus`

bradyallenjohnson commented:
> It looks like the output is the same as:
> 
> ```
> ipsec connectionstatus | grep '[.][.][.]'
> ```
> 
> I wonder what we would think of displaying if the command was called `ipsec routestatus`

I show the output in this email:
https://lists.libreswan.org/pipermail/swan-dev/2023-October/004654.html

paulwouters commented:
Personally, I strongly prefer the first two things are source/mask <=> dest/mask,
as that is the core info people usually want to see/confirm. Having this
split in two makes this much harder to read. If you want to avoid adding
up the counters, perhaps this format could be considered:

     000 172.16.20.0/24 <=> 172.16.10.0/24 from 172.22.18.102 to 172.22.18.101 (2KiB/1KiB) vpnclient.gwn02.xyz.com, reqid=16388

Thanks for adding reqid, that is very useful.
Perhaps peer ID should be the last entry and reqid should come first, as those peer IDs can be very long when certificates are used.

cagney commented:
Can the addition of `reqid` to `ipsec connectionstatus` please be moved to a separate patch - it will make a mess of testing so better to isolate it.

bleve commented:
Now handling multiple traffic-selectors is problematic. Current listing:

```
000 193.65.3.96/27 @ 193.65.2.90 (5KiB in)  <==>  188.127.201.224/27 @ 188.127.201.217 (4KiB in)        foo-bar-v4, reqid=16440
000 193.65.3.96/27 @ 193.65.2.90 (5KiB in)  <==>  172.23.128.0/23 @ 188.127.201.217 (4KiB in)   foo-bar-v4, reqid=16440
000 172.23.0.0/17 @ 193.65.2.90 (5KiB in)  <==>  188.127.201.224/27 @ 188.127.201.217 (4KiB in) foo-bar-v4, reqid=16440
000 172.23.0.0/17 @ 193.65.2.90 (5KiB in)  <==>  172.23.128.0/23 @ 188.127.201.217 (4KiB in)    foo-bar-v4, reqid=16440

```

I suggest something like this, Output changing also other direction traffic to out.

`
000 193.65.3.96/27,172.23.0.0/17 @ 193.65.2.90 (5KiB in)  <==>  188.127.201.224/27,172.23.128.0/23 @ 188.127.201.217 (4KiB out) foo-bar-v4, reqid=16440
`

letoams commented:
On Thu, 2 Nov 2023, cagney wrote:

> Subject: Re: [libreswan/libreswan] Changes to restore ipsec show command (PR

> Can the addition of reqid to ipsec connectionstatus please be moved to a separate patch - it will make a mess of
> testing so better to isolate it.

We didn't use "ipsec show" in tests anymore. We had removed it. I don't
think any of this would affect any tests?

letoams commented:
On Thu, 2 Nov 2023, Tuomo Soini wrote:

> Now handling multiple traffic-selectors is problematic. Current listing:
> 
> 000 193.65.3.96/27 @ 193.65.2.90 (5KiB in)  <==>  188.127.201.224/27 @ 188.127.201.217 (4KiB in)        foo-bar-v4,
>  reqid=16440
> 000 193.65.3.96/27 @ 193.65.2.90 (5KiB in)  <==>  172.23.128.0/23 @ 188.127.201.217 (4KiB in)   foo-bar-v4, reqid=1
> 6440
> 000 172.23.0.0/17 @ 193.65.2.90 (5KiB in)  <==>  188.127.201.224/27 @ 188.127.201.217 (4KiB in) foo-bar-v4, reqid=1
> 6440
> 000 172.23.0.0/17 @ 193.65.2.90 (5KiB in)  <==>  172.23.128.0/23 @ 188.127.201.217 (4KiB in)    foo-bar-v4, reqid=1
> 6440

Other than traffic counters, I think this is good.

> 000 193.65.3.96/27,172.23.0.0/17 @ 193.65.2.90 (5KiB in) <==> 188.127.201.224/27,172.23.128.0/23 @ 188.127.201.217
> (4KiB out) foo-bar-v4, reqid=16440

With more subnets this again only becomes machine parseable to read. The
idea of "ipsec show" is that humans can read it easilly.

cagney commented:
Once the command's name is settled we can move the code to `whack_<command>.[hc]` (this can happen post merge).

cagney commented:
Can I suggest dropping the header / footer - keep it brief.

Was each line going to be prefixed with the connection name?  Without this it is hard to tie the information back to the configuration.

Can I suggest including the routing so it is clear which are established, or are only established connections shown like for `ipsec trafficstatus`.

Is the name settled? `ipsec briefconnectionstatus` is a mouth full (I suggested routestatus before).

Having one SPD per line will make it easier to understand (but confusing when it comes to traffic counts as they are not per line), think about:
```
"my-connection": 192.0.1.23 ... 192.0.1.45 4kb in 8kb out routed-tunnel
"my-connection":   192.0.2.0/24 === 192.0.3.0/24 reqid 1
"my-connection":   1::/64 === 2:/64 reqid 2
```

letoams commented:
On Thu, 2 Nov 2023, cagney wrote:

> Can I suggest dropping the header / footer - keep it brief.

Yes.

> Was each line going to be prefixed with the connection name? Without this it is hard to tie the information back to
> the configuration.

I hope not. I think the idea is to closely keep with the old "ipsec
eroute" quick overview idea.

> Can I suggest including the routing so it is clear which are established, or are only established connections shown
> like for ipsec trafficstatus.

I think only established ones are?

> Is the name settled? ipsec briefconnectionstatus is a mouth full (I suggested routestatus before).

"ipsec eroute" ? :)

cagney commented:
> @cagney wrote
> > Was each line going to be prefixed with the connection name? Without this it is hard to tie the information back to the configuration.

@letoams wrote
> I hope not. I think the idea is to closely keep with the old "ipsec eroute" quick overview idea.

Then I'm struggling to see the advantage of this vs `ipsec show` and the host specific equivalents `ip xfrm`, `setkey`, et.al.

When I run `ipsec trafficstatus` and see a problem, I can follow it with `ipsec connectionstatus CONNECTION` (#1308) and examine the details.  Here I can't.

bradyallenjohnson commented:
> > @cagney wrote
> > > Was each line going to be prefixed with the connection name? Without this it is hard to tie the information back to the configuration.
> 
> @letoams wrote
> 
> > I hope not. I think the idea is to closely keep with the old "ipsec eroute" quick overview idea.
> 
> Then I'm struggling to see the advantage of this vs `ipsec show` and the host specific equivalents `ip xfrm`, `setkey`, et.al.
> 
> When I run `ipsec trafficstatus` and see a problem, I can follow it with `ipsec connectionstatus CONNECTION` (#1308) and examine the details. Here I can't.

Yes, the configuration name is indeed displayed in "ipsec connectionstatus" and in the new "ipsec briefconnectionstatus" command.
In this output, the configuration name (which is also the connection name) is vpnclient.gwn02.xyz.com:

000 [172.16.20.0/24](http://172.16.20.0/24) @ 172.22.18.102 (2KiB in)  <==>  [172.16.10.0/24](http://172.16.10.0/24) @ 172.22.18.101 (1KiB in) vpnclient.gwn02.xyz.com, reqid=16388

cagney commented:
@bradyallenjohnson wrote
> Yes, the configuration name is indeed displayed [...] In this output, the configuration name (which is also the connection name) is vpnclient.gwn02.xyz.com:
```
000 [172.16.20.0/24](http://172.16.20.0/24) @ 172.22.18.102 (2KiB in) <==> [172.16.10.0/24](http://172.16.10.0/24) @ 172.22.18.101 (1KiB in) vpnclient.gwn02.xyz.com, reqid=16388
```
ah, so that is what that is, good!  I had assumed it was some DNS info or some such and ignored it.  It should be displayed using jam_connection(), I added a note.

bradyallenjohnson commented on programs/pluto/whack_connectionstatus.c:
```diff
@@ -325,7 +329,7 @@ static void show_one_spd(struct show *s,
 		jam_string(buf, " their_ip=");
 		OPT_HOST(their_sourceip);
-		jam_string(buf, ";");
+		jam(buf, "; reqid=%u;", c->child.reqid);
```
Ok, I'll remove this from this PR.

bradyallenjohnson commented on programs/pluto/whack_connectionstatus.c:
```diff
@@ -934,3 +938,156 @@ void whack_connectionstatus(const struct whack_message *m, struct show *s)
+		jam_humber_uintmax(buf, "/", get_child_bytes(c, DIRECTION_OUTBOUND), "B)");
+
+		// name and reqid
+		jam(buf, "\t"PRI_CONNECTION", reqid=%u", c->name, instance, c->child.reqid);
```
Ok, I will change this.

bradyallenjohnson commented on programs/pluto/whack_connectionstatus.h:
```diff
@@ -20,6 +20,7 @@ struct whack_message;
 struct show;
 
 void whack_connectionstatus(const struct whack_message *m, struct show *s);
+void whack_brief_connectionstatus(const struct whack_message *m, struct show *s);
```
Ok

bradyallenjohnson commented on programs/pluto/whack_connectionstatus.c:
```diff
@@ -46,7 +46,11 @@
+#include "kernel.h"		/* for enum direction */
+
+/* Passed in to jam_end_client() */
+static const char END_SEPARATOR[] = "===";
```
The END_SEPARATOR is indeed used in the calls to jam_end_client(). The jam_end_client() function is called with the separator parameter set to either NULL or END_SEPARATOR.

cagney replied:
Yes, it's treated like a boolean though
```
	if (left_right == RIGHT_END && separator != NULL) {
		jam_string(buf, "===");
	}
```

bradyallenjohnson replied:
Oh, I see, I'll fix that, nice catch.

bleve commented:
I don't see any need for output prefix and suffix parts, I suggest removing those.

bleve commented:
And I still don't like this:
```

000 193.65.3.96/27 <==> 188.127.201.224/27	from 193.65.2.90 to 188.127.201.217 (119KiB/112KiB)	"foo-bar-v4", reqid=16440
000 193.65.3.96/27 <==> 172.23.128.0/23	from 193.65.2.90 to 188.127.201.217 (119KiB/112KiB)	"foo-bar-v4", reqid=16440
000 172.23.0.0/17 <==> 188.127.201.224/27	from 193.65.2.90 to 188.127.201.217 (119KiB/112KiB)	"foo-bar-v4", reqid=16440
000 172.23.0.0/17 <==> 172.23.128.0/23	from 193.65.2.90 to 188.127.201.217 (119KiB/112KiB)	"foo-bar-v4", reqid=16440

```

When we could do this all in one line.

```
000 193.65.3.96/27,172.23.0.0/17 <==> 188.127.201.224/27,172.23.128.0/23        from 193.65.2.90 to 188.127.201.217 (119KiB/112KiB)     "foo-bar-v4", reqid=16440
```

bradyallenjohnson commented:
> I don't see any need for output prefix and suffix parts, I suggest removing those.

Which prefix/suffix parts are you referring to?

bleve commented:
> > I don't see any need for output prefix and suffix parts, I suggest removing those.
> 
> Which prefix/suffix parts are you referring to?

```
000 Connection list:
000  
```

And

```
000  
000 Total IPsec connections: loaded 38, active 19
```

bradyallenjohnson commented:
> > > I don't see any need for output prefix and suffix parts, I suggest removing those.
> > 
> > 
> > Which prefix/suffix parts are you referring to?
> 
> ```
> 000 Connection list:
> 000  
> ```
> 
> And
> 
> ```
> 000  
> 000 Total IPsec connections: loaded 38, active 19
> ```

The final line seems very useful to me. If there are more than 4 or 5 connections, I would like to have a simple count of the number of connections.

bradyallenjohnson commented:
> And I still don't like this:
> 
> ```
> 
> 000 193.65.3.96/27 <==> 188.127.201.224/27	from 193.65.2.90 to 188.127.201.217 (119KiB/112KiB)	"foo-bar-v4", reqid=16440
> 000 193.65.3.96/27 <==> 172.23.128.0/23	from 193.65.2.90 to 188.127.201.217 (119KiB/112KiB)	"foo-bar-v4", reqid=16440
> 000 172.23.0.0/17 <==> 188.127.201.224/27	from 193.65.2.90 to 188.127.201.217 (119KiB/112KiB)	"foo-bar-v4", reqid=16440
> 000 172.23.0.0/17 <==> 172.23.128.0/23	from 193.65.2.90 to 188.127.201.217 (119KiB/112KiB)	"foo-bar-v4", reqid=16440
> ```
> 
> When we could do this all in one line.
> 
> ```
> 000 193.65.3.96/27,172.23.0.0/17 <==> 188.127.201.224/27,172.23.128.0/23        from 193.65.2.90 to 188.127.201.217 (119KiB/112KiB)     "foo-bar-v4", reqid=16440
> ```

I disagree. The second output is much harder to read.

bleve commented:
We have `ipsec briefstatus` for that.

cagney commented:
Having one SPD per line is consistent with the old `ipsec show` command and makes parsing the output using scripts much easier.  This isn't a log file where we're trying to squeeze as much as possible on a single line.

cagney commented:
I don't think the output should be decorated with the header/footer as it, again, makes it more consistent with the old `ipsec show`, however ...

`ipsec briefconnectionstatus ; ipsec briefstatus` is *not* equivalent to including the header footer but `ipsec whack --briefconnectionstatus --briefstatus` kind of is (yea, long standing quirk in how this works).

cagney commented:
@bradyallenjohnson  I still see
```
@@ -325,7 +329,7 @@ static void show_one_spd(struct show *s,
                ip_address their_sourceip = spd_end_sourceip(spd->remote);
                jam_string(buf, " their_ip=");
                OPT_HOST(their_sourceip);
-               jam_string(buf, ";");
+               jam(buf, "; reqid=%u;", c->child.reqid);
 #undef OPT_HOST
        }
 
```

bradyallenjohnson commented:
> jam_string(buf, ";");

Rebased and fixed in the latest push.

cagney commented:
I've merged and shuffled things so it is easier to look at and play with.  And added your (C).

What happens in transport mode when host==client?  It looks like jam_end_client() prints nothing.
We still lack documentation, tests, and consensus on the final output.

bradyallenjohnson commented:
> I've merged and shuffled things so it is easier to look at and play with. And added your (C).
> 
> What happens in transport mode when host==client? It looks like jam_end_client() prints nothing. We still lack documentation, tests, and consensus on the final output.

There are 3 cases affected by this: transport mode when host==client, tunnel mode with host-to-subnet, and tunnel mode with subnet-to-host.

I propose the following output for these 3 cases:

transport mode when host == client for both the local and remote:
000 host <==> host from 172.22.18.102 to 172.22.18.101 (0B/0B) "gwn02_transport_tun", reqid=16388

tunnel mode with host-to-subnet:
000 host <==> 172.16.10.0/24  from 172.22.18.102 to 172.22.18.101 (0B/0B) "gwn02_transport_tun", reqid=16388

tunnel mode with subnet-to-host:
000 172.16.20.0/24 <==> host  from 172.22.18.102 to 172.22.18.101 (0B/0B) "gwn02_transport_tun", reqid=16388

letoams commented:
On Mon, 20 Nov 2023, Brady Johnson wrote:

> There are 3 cases affected by this: transport mode when host==client, tunnel mode with host-to-subnet,
> and tunnel mode with subnet-to-host.
> 
> I propose the following output for these 3 cases:
> 
> transport mode when host == client for both the local and remote:
> 000 host <==> host from 172.22.18.102 to 172.22.18.101 (0B/0B) "gwn02_transport_tun", reqid=16388
> 
> tunnel mode with host-to-subnet:
> 000 host <==> 172.16.10.0/24 from 172.22.18.102 to 172.22.18.101 (0B/0B) "gwn02_transport_tun",
> reqid=16388
> 
> tunnel mode with subnet-to-host:
> 000 172.16.20.0/24 <==> host from 172.22.18.102 to 172.22.18.101 (0B/0B) "gwn02_transport_tun",
> reqid=16388

Instead of the string "host" I would just repeat the IP/mask of the
host instead. This will make it easier on tools parsing the output
automatically - they wouldn't really need to know if it was tunnel
or transport or net-net or host-net/net-host.

Paul

bradyallenjohnson commented:
Ok, Paul, I'll do that and push the changes.

bradyallenjohnson committed:
Changes to restore ipsec show command

- Now called "ipsec briefconnectionstatus", and
  is a simpler version "ipsec connectionstatus"
- Now implemented in c-code in the whack binary

Signed-off-by: Brady Johnson <bradyallenjohnson@gmail.com>

bradyallenjohnson commented:
I just pushed changes to handle the 4 use cases of host2host (transport mode), host2subnet, subnet2host, and subnet2subnet with the output suggested by @letoams as follows:

000 172.22.18.102 <==> 172.22.18.101	from 172.22.18.102 to 172.22.18.101 (0B/0B)	"gwn02_transport_tun", reqid=16388

000 172.22.18.102 <==> 172.16.10.0/24	from 172.22.18.102 to 172.22.18.101 (0B/0B)	"gwn02_host2subnet_tun", reqid=16392

000 172.16.20.0/24 <==> 172.22.18.101	from 172.22.18.102 to 172.22.18.101 (0B/0B)	"gwn02_subnet2host_tun", reqid=16388

000 172.16.20.0/24 <==> 172.16.10.0/24	from 172.22.18.102 to 172.22.18.101 (0B/0B)	"vpnclient.gwn02.xyz.com", reqid=16396

@cagney Can you point out where the documentation for this is so I can update it. Also, for the tests, is the libreswan/testing/pluto/whack-02-globalstatus test a good example to follow?

cagney commented:
I merged.

I do wonder if the jam_end_child() calls can be reduced to jam_selector() and jam_end_host() to jam_address() when showing the SPD.  BTW, like how it shows `172.22.18.102` (i.e., address) and not `172.22.18.102/32` (i.e., selector).

For documentation, it's in programs/ipsec/ipsec_<cmd>.8.xml, I'm not sure how @bleve is keeping it all co-ordinated.

For tests, tricky, I've never found a good basic template (the basic tests often aren't that basic).  Perhaps `testing/pluto/ikev2-05-basic-psk`  Since you've figured out the configs, you could tweak that to up/show/down each combination.

cagney commented:
Moved to 5.1 as not a blocker.  Since `whack_briefconnectionstatus.[hc]` is nice an isolated we can, in truth, still pull stuff into 5.0

bleve commented:
This pull request was already merged in, closing. For more changes, create a new one.
